src/toil/jobStores/aws/jobStore.py

```diff
@@ -1186,24 +1185,31 @@ def readFrom(self, readable):
                         # We will compute a checksum
                         hasher = info._start_checksum()
                         info._update_checksum(hasher, buf)
-                    
-                        headers = info._s3EncryptionHeaders()
-                        for attempt in retry_s3():
-                            with attempt:
-                                log.debug('Starting multipart upload')
-                                upload = store.filesBucket.initiate_multipart_upload(
-                                    key_name=compat_bytes(info.fileID),
-                                    headers=headers)
+
+                        # Below is using boto3's low level API for uploading parts manually. Note that
+                        # client.upload_file() can automatically upload with multiple parts, though I
+                        # don't want to mess with the checksum logic.
+
+                        log.debug('Starting multipart upload')
+                        # low-level clients are thread safe
+                        client = store.s3_client
+                        bucket_name = store.filesBucket.name
+                        headerArgs = info._s3EncryptionArgs()
+
+                        upload = client.create_multipart_upload(Bucket=bucket_name,
+                                                                Key=compat_bytes(info.fileID), **headerArgs)
+                        uploadId = upload['UploadId']
+                        parts = []
+
                         try:
                             for part_num in itertools.count():
-                                for attempt in retry_s3():
-                                    with attempt:
-                                        log.debug('Uploading part %d of %d bytes', part_num + 1, len(buf))
-                                        upload.upload_part_from_file(fp=BytesIO(buf),
-                                                                     # part numbers are 1-based
-                                                                     part_num=part_num + 1,
-                                                                     headers=headers)
-                                
+                                log.debug('Uploading part %d of %d bytes', part_num + 1, len(buf))
+                                part = client.upload_part(Bucket=bucket_name, Key=compat_bytes(info.fileID),
```

Comment:
I would keep the retries here for 404, 500, 502, 503, 504.  I would also a comment `TODO` to include the `Content-MD5` header in the future: https://boto3.amazonaws.com/v1/documentation/api/latest/reference/services/s3.html#S3.Client.complete_multipart_upload